### PR TITLE
New mitohifi/findmitoreference module

### LINKED ISSUE #2571: new module: MitoHiFi
### Is there an existing module for this?

- [X] I have searched for the existing module

### Is there an open PR for this?

- [X] I have searched for existing PRs

### Is there an open issue for this?

- [X] I have searched for existing issues

### Are you going to work on this?

- [X] If I'm planning to work on this module, I added myself to the `Assignees` to facilitate tracking who is working on the module

Tool: https://github.com/marcelauliano/MitoHiFi

### PR DESCRIPTION
**Pull request by verku**
PR checklist

Working on #2571 (need mitohifi/mitohifi to close the issue)

- [x]  This comment contains a description of changes (with reason).
- [x]  If you've fixed a bug or added code that should be tested, add tests!
- [x]  If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ]  If necessary, include test data in your PR.
- [x]  Remove all TODO statements.
- [x]  Emit the versions.yml file.
- [x]  Follow the naming conventions.
- [x]  Follow the parameters requirements.
- [x]  Follow the input/output options guidelines.
- [x]  Add a resource label
- [ ]  Use BioConda and BioContainers if possible to fulfil software requirements.
- [x] Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [x]  PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware
    - [x]  PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware
    - [ ]  PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd --git-aware  

- Added a module to run the tool findMitoReference.py from MitoHiFi
- The MitoHifi module itself is still missing

I have a few open questions/issues that I'm adding below.

### CODE REVIEW

File: modules/nf-core/mitohifi/findmitoreference/main.nf [hunk 1]
```diff
@@ -0,0 +1,36 @@
+process MITOHIFI_FINDMITOREFERENCE {
+    tag '$species'
+    label 'process_low'
+
+    // MitoHifi does not exist as conda package
+    // The MitoHiFi image is only available on Dockerhub
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
```
Comment (verku) [inline modules/nf-core/mitohifi/findmitoreference/main.nf:7]: What does this line of code exactly do?

File: tests/modules/nf-core/mitohifi/findmitoreference/test.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+## TODO nf-core: Please run the following command to build this file:
+#                nf-core modules create-test-yml mitohifi/findmitoreference
```
Comment (verku) [inline tests/modules/nf-core/mitohifi/findmitoreference/test.yml:2]: This command fails with the following error message:

`INFO     Running 'mitohifi/findmitoreference' test with command:                                                                                                                 
         nextflow run ./tests/modules/nf-core/mitohifi/findmitoreference -entry test_mitohifi_findmitoreference -c ./tests/config/nextflow.config -c                             
         ./tests/modules/nf-core/mitohifi/findmitoreference/nextflow.config --outdir /var/folders/p0/6z00kpv16qbf_bt52y4zz2kc0000gp/T/tmp9zlc7am9 -work-dir                      
         /var/folders/p0/6z00kpv16qbf_bt52y4zz2kc0000gp/T/tmp2hul0uxf                                                                                                            
CRITICAL Error running test workflow (exit code 1)                                                                                                                               
         N E X T F L O W  ~  version 22.10.6                                                                                                                                     
         Launching `./tests/modules/nf-core/mitohifi/findmitoreference/main.nf` [prickly_curran] DSL2 - revision: f478a3f645                                                     
         Module compilation error                                                                                                                                                
         - file :                                                                                                                                                                
         /Users/verenakutschera/nobackup/misc/hackathons/202303_nf-core/modules/tests/modules/nf-core/mitohifi/findmitoreference/../../../../../modules/nf-core/mitohifi/findmito
         reference/main.nf                                                                                                                                                       
         - cause: Unexpected input: '{' @ line 1, column 36.                                                                                                                     
            process MITOHIFI_FINDMITOREFERENCE {                                                                                                                                 
                                               ^                                                                                                                                 
                                                                                                                                                                                 
         1 error`

File: modules/nf-core/mitohifi/findmitoreference/main.nf [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        'docker://biocontainers/mitohifi:2.2_cv1' }"
+
+    input:
+    val species
+    val email
+    val min_length
```
Comment (nvnieuwk) [inline modules/nf-core/mitohifi/findmitoreference/main.nf:13]: Are these values all mandatory?

File: modules/nf-core/mitohifi/findmitoreference/main.nf [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        --email $email \\
+        --min_length $min_length \\
+        --outfolder .
+
+     cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        mitohifi: \$( mitohifi.py --version )
+    END_VERSIONS
+    """
+}
```
Comment (nvnieuwk) [inline modules/nf-core/mitohifi/findmitoreference/main.nf:36]: Can you also add a stub section here? See [this](https://github.com/nf-core/modules/blob/master/modules/nf-core/bcftools/filter/main.nf) for an example

File: modules/nf-core/mitohifi/mitohifi/main.nf
Comment (nvnieuwk) [inline modules/nf-core/mitohifi/mitohifi/main.nf:1]: Can you make a separate PR for this module? It's easier to divide the reviewing process :)

File: tests/modules/nf-core/mitohifi/findmitoreference/main.nf [hunk 5]
```diff
@@ -0,0 +1,14 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { MITOHIFI_FINDMITOREFERENCE } from '../../../../../modules/nf-core/mitohifi/findmitoreference/main.nf'
+
+workflow test_mitohifi_findmitoreference {
+    
+    species = "Homo sapiens"
+    email = "verena.kutschera@scilifelab.se"
```
Comment (nvnieuwk) [inline tests/modules/nf-core/mitohifi/findmitoreference/main.nf:10]: I wouldn't put your personal e-mail here, unless you want to receive constant e-mails when someone reruns the tests from this module :)

**Review by nvnieuwk — CHANGES_REQUESTED** [review]

File: modules/nf-core/mitohifi/findmitoreference/main.nf [hunk 6]
```diff
@@ -0,0 +1,36 @@
+process MITOHIFI_FINDMITOREFERENCE {
+    tag '$species'
+    label 'process_low'
+
+    // MitoHifi does not exist as conda package
```
[the remark this answers is not in the document]
Comment (verku) [reply modules/nf-core/mitohifi/findmitoreference/main.nf:5]: I've never done that before, so I'm not sure! 

However, the developers of MitoHifi write that one of the dependencies uses a different python version than the others, which is why they suggest either using their container or a conda environment+manual installation of that tool: https://github.com/marcelauliano/MitoHiFi#21-manual-installation---dependencies

File: modules/nf-core/mitohifi/findmitoreference/main.nf (hunk 3 above)
Comment (verku) [reply modules/nf-core/mitohifi/findmitoreference/main.nf:13]: Yes, they are all mandatory

File: modules/nf-core/mitohifi/findmitoreference/main.nf (hunk 6 above)
[the remark this answers is not in the document]
Comment (nvnieuwk) [reply modules/nf-core/mitohifi/findmitoreference/main.nf:5]: You can always ask for help on the nf-core slack or the Gitter of bioconda, I'm not that experienced in this myself

File: tests/modules/nf-core/mitohifi/findmitoreference/main.nf (hunk 5 above)
Comment (verku) [reply tests/modules/nf-core/mitohifi/findmitoreference/main.nf:10]: Yes, good point... ;-)

Comment (verku) [reply tests/modules/nf-core/mitohifi/findmitoreference/main.nf:10]: Do you have a suggestion for an email address I could enter instead? Like some kind of dummy email address?

Comment (nvnieuwk) [reply tests/modules/nf-core/mitohifi/findmitoreference/main.nf:10]: Go wild ;) but make sure it's not taken already so you won't spam anyone

File: tests/modules/nf-core/mitohifi/findmitoreference/test.yml (hunk 2 above)
Comment (verku) [reply tests/modules/nf-core/mitohifi/findmitoreference/test.yml]: This has finally worked! :-)

File: modules/nf-core/mitohifi/findmitoreference/main.nf (hunk 6 above)
[the remark this answers is not in the document]
Comment (verku) [reply modules/nf-core/mitohifi/findmitoreference/main.nf:5]: I got in touch with some bioconda people! Unfortunately, MitoHifi can't be made available as bioconda package or biocontainer. It is essentially a mini-pipeline written in Python, and the tools it runs need different versions of Python, which can't be installed in the same Conda environment. That’s why there is instead a Docker container.

I assume this rules out having it as a nf-core module?

File: modules/nf-core/mitohifi/findmitoreference/main.nf [hunk 7]
```diff
@@ -30,7 +30,15 @@ process MITOHIFI_FINDMITOREFERENCE {
...
-        mitohifi: \$( mitohifi.py --version )
+        mitohifi: \$( mitohifi.py --version 2>&1 | head -n1 | sed 's/^.*MitoHiFi //; s/ .*\$//' )
+    END_VERSIONS
+    """
+
+    stub:
+    """
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        mitohifi: \$( mitohifi.py --version 2>&1 | head -n1 | sed 's/^.*MitoHiFi //; s/ .*\$//' )
```
Comment (nvnieuwk) [inline modules/nf-core/mitohifi/findmitoreference/main.nf:42]:
```suggestion
    stub:
    """
    touch ${prefix}.fasta
    touch ${prefix}.gb

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        mitohifi: \$( mitohifi.py --version 2>&1 | head -n1 | sed 's/^.*MitoHiFi //; s/ .*\$//' )
```

The goal of a stub run is to emulate the actual process so the input and output files should be as close as possible to the real input and output. Creating empty files is sufficient in this case. (The stub will throw an error if the fasta and gb files aren't there)

Comment (verku) [reply modules/nf-core/mitohifi/findmitoreference/main.nf:42]: Aaah, that makes sense! Thanks!

File: modules/nf-core/mitohifi/findmitoreference/main.nf (hunk 6 above)
[the remark this answers is not in the document]
Comment (nvnieuwk) [reply modules/nf-core/mitohifi/findmitoreference/main.nf:5]: Hi, thanks for asking! I think it's fine how it's done now, but I'll have to ask for a second opinion here :) (@jfy133, @maxulysse, @ewels any opinions here?)

File: tests/modules/nf-core/mitohifi/findmitoreference/test.yml (hunk 2 above)
Comment (nvnieuwk) [reply tests/modules/nf-core/mitohifi/findmitoreference/test.yml]: Awesome!

File: modules/nf-core/mitohifi/findmitoreference/main.nf (hunk 7 above)
Comment (nvnieuwk) [reply modules/nf-core/mitohifi/findmitoreference/main.nf]: Oh you should also add the creation of the `prefix` variable to the stub section :)

Comment (nvnieuwk) [reply modules/nf-core/mitohifi/findmitoreference/main.nf]: Oh right I forgot about that, my bad 😁. You can just use dummy file names in this case :)

Comment (verku) [reply modules/nf-core/mitohifi/findmitoreference/main.nf]: No worries, I'll do that! :-)

File: modules/nf-core/mitohifi/findmitoreference/main.nf (hunk 6 above)
[the remark this answers is not in the document]
Comment (jfy133) [reply modules/nf-core/mitohifi/findmitoreference/main.nf:5]: I think conceptually it _may_ be OK, although I'm wary of non-standard container registries. That said github container registry should be pretty stable I would assume?

@FriederikeHanssen I vaguely remember you brought up a non-conda package at some point (maybe wrong...)

[the remark this answers is not in the document]
Comment (verku) [reply modules/nf-core/mitohifi/findmitoreference/main.nf:5]: The biocontainers image contains version 2.14.2, the github image has version 3.0.0 - but maybe biocontainers is more stable...

**Review by nvnieuwk — APPROVED** [review]

File: modules/nf-core/mitohifi/findmitoreference/main.nf (hunk 6 above)
[the remark this answers is not in the document]
Comment (muffato) [reply modules/nf-core/mitohifi/findmitoreference/main.nf:5]: Hiya. I work with the MitoHifi people. 
I'll get the latest Dockerfile into BioContainers because there's a lot of new stuff in the new version. Can you hold on until that's done ?

[the remark this answers is not in the document]
Comment (muffato) [reply modules/nf-core/mitohifi/findmitoreference/main.nf:5]: https://github.com/BioContainers/containers/pull/509 🤞🏼

[the remark this answers is not in the document]
Comment (muffato) [reply modules/nf-core/mitohifi/findmitoreference/main.nf:5]: @verku : the latest container is now `biocontainers/mitohifi:3.0.0_cv1`

**Review by muffato — COMMENTED** [review]
The fact that Conda is not supported should be listed at https://github.com/nf-core/modules/blob/master/.github/workflows/pytest-workflow.yml#L42 . This way you'll have a clean sheet for the tests

File: modules/nf-core/mitohifi/findmitoreference/main.nf (hunk 6 above)
[the remark this answers is not in the document]
Comment (verku) [reply modules/nf-core/mitohifi/findmitoreference/main.nf:5]: Thank you, this is great news! I'll update the module accordingly.

**Comment by verku** [comment]
> The fact that Conda is not supported should be listed at https://github.com/nf-core/modules/blob/master/.github/workflows/pytest-workflow.yml#L42 . This way you'll have a clean sheet for the tests

Good to know that this is possible, I'll do that!

File: tests/modules/nf-core/mitohifi/findmitoreference/test.yml [hunk 8]
```diff
@@ -0,0 +1,11 @@
+- name: mitohifi findmitoreference test_mitohifi_findmitoreference
+  command: nextflow run ./tests/modules/nf-core/mitohifi/findmitoreference -entry test_mitohifi_findmitoreference -c ./tests/config/nextflow.config -c ./tests/modules/nf-core/mitohifi/findmitoreference/nextflow.config
+  tags:
+    - mitohifi/findmitoreference
+    - mitohifi
+  files:
+    - path: output/mitohifi/NC_072273.1.fasta
+      md5sum: 14c9906c7eccc0e57482112a452a031a
+    - path: output/mitohifi/NC_072273.1.gb
+      md5sum: 52d0342e6a9d0c3de076e68ef800eea9
```
Comment (muffato) [inline tests/modules/nf-core/mitohifi/findmitoreference/test.yml:10]: The record got updated on Monday ! Here is the new MD5
```suggestion
      md5sum: 9063efc0ee83569566592f0198aa4eb5
```
